### DISCUSSION TITLE
What is the timeline of order execution & claiming breach and bankrupt?

### DISCUSSION THREAD

**Discussion by rezunli96** [yasserfarouk/scml#21] [standard-qa] [answered]
Hello I have some confusion about when will one get claimed breach/bankrupt in std/col. In the example of the document it seems the agent will first get breached due to some sale order, and then bankrupt due to some buy order. But this relies on the execution process of order right?

Suppose say, at day x the agent are supposed to execute sale order A, buy order B and sale order C. If the timeline is B-A-C, then agent will get bankrupted at B. But suppose if the order is A-C-B, then the agent may not got bankrupted. So is it like the contract is executed one at a time or in a oneshot manner? If it is in the previous case, is there any prioritization among the orders (like sale orders are always before buy ones)?

I guess it is in start_contract_execution(..) of scml2020/world.py that the above process happen right? From my first reading it is handling one contract at a time?

And I believe it may be more or less related to what described in the last bullet of Section 7. What does it imply by "Nevertheless, it can use input products in production on the same day they are bought. Similarly, it can use funds earned from sales for production on the same day products are sold."?

Thanks

**Comment by yasserfarouk** [comment, marked as answer]
> But this relies on the execution process of order right?

Yes it does. Contracts are executed in the order of products (i.e. from raw material to final product). An agent cannot have sell and buy contracts for the same product. The uncertainty (randomness) in execution order is within contracts for the same product. These are completely random.

The relevant code is here:

> is there any prioritization among the orders (like sale orders are always before buy ones)?

For any specific agent, buy ones come before sell ones.

> I guess it is in start_contract_execution(..) of scml2020/world.py that the above process happen right? From my first reading it is handling one contract at a time?

It is in `order_contracts_for_execution` and yes it does.

```python
    def order_contracts_for_execution(
        self, contracts: Collection[Contract]
    ) -> Collection[Contract]:
        return sorted(contracts, key=lambda x: x.annotation["product"])
 ```

>  What does it imply by "Nevertheless, it can use input products in production on the same day they are bought. Similarly, it can use funds earned from sales for production on the same day products are sold."?

That is kind of different from but related to contract execution order. It is about the time needed for production. Time needed for production is zero and contracts are executed in the order of the product index. This implies that the input materials will all be available (if enough money is in the wallet) before needing to sell anything (that is the first part). I may need to go back to check the details of how the second point applies though.

**Reply by rezunli96** [reply]
So what could be the source of randomness in execution order within contracts for the same product?

**Reply by yasserfarouk** [reply]
They are run in the order they are concluded and the order in which they are concluded depends on the order in which negotiations are run which we explicitly randomize and the number of rounds each negotiation takes  plus the time spent in each round (when negotiations run in parallel) which introduce another source of indeterminacy
